proto/ask_plan.go

```diff
@@ -37,9 +46,224 @@ func (m *AskPlan) Validate() error {
 		return errors.New("RAM size is too low")
 	}
 
-	if len(m.GetResources().GetGPU().GetHashes()) > 0 && len(m.GetResources().GetGPU().GetIndexes()) > 0 {
+	return m.GetResources().GetGPU().Validate()
+}
+
+func NewEmptyAskPlanResources() *AskPlanResources {
+	res := &AskPlanResources{}
+	res.InitNilWithZero()
+	return res
+}
+
+func (m *AskPlanResources) InitNilWithZero() {
+	if m.CPU == nil {
+		m.CPU = &AskPlanCPU{}
+	}
+	if m.RAM == nil {
+		m.RAM = &AskPlanRAM{}
+	}
+	if m.RAM.Size == nil {
+		m.RAM.Size = &DataSize{}
+	}
+	if m.Storage == nil {
+		m.Storage = &AskPlanStorage{}
+	}
+	if m.Storage.Size == nil {
+		m.Storage.Size = &DataSize{}
+	}
+	if m.GPU == nil {
+		m.GPU = &AskPlanGPU{}
+	}
+	if m.GPU.Hashes == nil {
+		m.GPU.Hashes = []string{}
+	}
+	if m.GPU.Indexes == nil {
+		m.GPU.Indexes = []uint64{}
+	}
+	if m.Network == nil {
+		m.Network = &AskPlanNetwork{}
+	}
+	if m.Network.ThroughputOut == nil {
+		m.Network.ThroughputOut = &DataSizeRate{}
+	}
+	if m.Network.ThroughputIn == nil {
+		m.Network.ThroughputIn = &DataSizeRate{}
+	}
+}
+
+func (m *AskPlanResources) Add(resources *AskPlanResources) error {
+	m.InitNilWithZero()
+	if err := m.GetGPU().Add(resources.GetGPU()); err != nil {
+		return err
+	}
+	m.CPU.CorePercents += resources.GetCPU().CorePercents
+	m.RAM.Size.Bytes += resources.GetRAM().GetSize().GetBytes()
+	m.Storage.Size.Bytes += resources.GetStorage().GetSize().GetBytes()
+	m.Network.Incoming = m.GetNetwork().GetIncoming() && resources.GetNetwork().GetIncoming()
+	m.Network.Outbound = m.GetNetwork().GetOutbound() && resources.GetNetwork().GetOutbound()
+	m.Network.Overlay = m.GetNetwork().GetOverlay() && resources.GetNetwork().GetOverlay()
+	m.Network.ThroughputIn.BitsPerSecond += resources.GetNetwork().GetThroughputIn().GetBitsPerSecond()
+	m.Network.ThroughputOut.BitsPerSecond += resources.GetNetwork().GetThroughputOut().GetBitsPerSecond()
+	return nil
+}
+
+func (m *AskPlanResources) Sub(resources *AskPlanResources) error {
+	m.InitNilWithZero()
+	if ok, desc := m.Contains(resources); !ok {
+		return errors.New(desc)
+	}
+	m.CPU.CorePercents -= resources.GetCPU().GetCorePercents()
+	m.RAM.Size.Bytes -= resources.GetRAM().GetSize().GetBytes()
+	m.Storage.Size.Bytes -= resources.GetStorage().GetSize().GetBytes()
+	m.GPU.Sub(resources.GetGPU())
+	m.Network.ThroughputIn.BitsPerSecond -= resources.GetNetwork().GetThroughputIn().GetBitsPerSecond()
+	m.Network.ThroughputOut.BitsPerSecond -= resources.GetNetwork().GetThroughputOut().GetBitsPerSecond()
+	return nil
+}
+
+func (m *AskPlanResources) ToHostConfigResources(cgroupParent string) container.Resources {
+	//TODO: check and discuss
+	return container.Resources{
```

Comment:
What the fuck is this? Check the code you've removed.

Reply:
fixed

Reply:
Where?

Reply:
I mean I've put exactly the same code as it was there previously